// @vuesion/webpack/dist/config/server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const webpack = require("webpack");
const utils_1 = require("./utils");
const base_server_1 = require("./base-server");
const path_1 = require("@vuesion/utils/dist/path");
const randomGenerator_1 = require("@vuesion/utils/dist/randomGenerator");
const CopyWebpackPlugin = require('copy-webpack-plugin');
const StartServerPlugin = require('start-server-webpack-plugin');
exports.server = utils_1.merge(base_server_1.baseServer, {
    name: 'server',
    entry: [path_1.runtimeRoot('src/server/index')],
    output: {
        filename: 'server.js',
    },
    plugins: [
        new CopyWebpackPlugin([
            { from: path_1.runtimeRoot('src/static'), to: '../static' },
            { from: path_1.runtimeRoot('src/app/config/*.json'), to: 'app/config', flatten: true },
        ]),
    ],
});
if (utils_1.isDev) {
    exports.server = utils_1.merge(exports.server, {
        watch: true,
        entry: ['webpack/hot/poll?1000'],
        plugins: [
            new StartServerPlugin({
                name: 'server.js',
                nodeArgs: [`--inspect=${randomGenerator_1.getIntInRange(9000, 9999)}`],
            }),
            new webpack.HotModuleReplacementPlugin(),
        ],
    });
}
exports.default = exports.server;
