// @vuesion/service/dist/commands/test.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("../decorators/command");
let Test = class Test {
    async run(args) {
        process.env.NODE_ENV = 'test';
        const jest = require('jest');
        jest.run(args);
    }
};
Test = __decorate([
    command_1.Command({
        name: 'test',
        alias: 't',
        description: 'Run unit-tests with jest. All Jest CLI options are supported.',
        options: [{ flags: '-c, --coverage', description: 'Run tests with coverage.' }],
    })
], Test);
exports.Test = Test;
