// @vuesion/models/dist/JSONModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const prettier_1 = require("prettier");
const utils_1 = require("@vuesion/utils");
let prettierConfig;
try {
    prettierConfig = JSON.parse(fs.readFileSync(utils_1.runtimeRoot('.prettierrc')).toString());
}
catch (e) {
    prettierConfig = {
        singleQuote: true,
        bracketSpacing: true,
        arrowParens: 'always',
        trailingComma: 'all',
        tabWidth: 2,
        printWidth: 120,
        endOfLine: 'lf',
    };
}
class JSONModel {
    constructor(path) {
        this.path = null;
        this.model = null;
        this.path = path;
        this.load();
    }
    load() {
        if (fs.existsSync(this.path)) {
            try {
                this.model = JSON.parse(fs.readFileSync(this.path).toString());
            }
            catch (e) {
                utils_1.logError(e);
            }
        }
    }
    save(prettier = true) {
        let jsonString = '';
        try {
            jsonString = JSON.stringify(this.model, null, prettierConfig.tabWidth) + '\n';
            const data = prettier
                ? prettier_1.format(jsonString, {
                    ...prettierConfig,
                    parser: 'json',
                })
                : jsonString;
            fs.writeFileSync(this.path, data);
        }
        catch (e) {
            utils_1.logError(e);
        }
    }
}
exports.JSONModel = JSONModel;
