// @vuesion/models/dist/VuesionConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const JSONModel_1 = require("./JSONModel");
const utils_1 = require("@vuesion/utils");
const ramda_1 = require("ramda");
class Model extends JSONModel_1.JSONModel {
    constructor() {
        super(utils_1.runtimeRoot('.vuesion/config.json'));
        if (this.model !== null) {
            this.clean = this.model.clean;
            this.currentVersion = this.model.currentVersion;
            this.devServer = this.model.devServer;
            this.generators = this.model.generators;
            this.i18n = this.model.i18n;
            this.prettier = this.model.prettier;
            this.spa = this.model.spa;
            this.webpack = this.model.webpack;
        }
    }
    getWebpackAliases() {
        return ramda_1.pathOr(null, ['aliases'], this.webpack);
    }
    updateCurrentVersion(currentVersion) {
        this.model.currentVersion = currentVersion;
        this.save();
    }
    getCustomWebpackConfig(config, target) {
        return require(utils_1.runtimeRoot('.vuesion/webpack.config'))(config, target);
    }
}
exports.VuesionConfig = new Model();
