// @vuesion/models/dist/VuesionPackage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const JSONModel_1 = require("./JSONModel");
const utils_1 = require("@vuesion/utils");
const VuesionConfig_1 = require("./VuesionConfig");
class Model extends JSONModel_1.JSONModel {
    constructor() {
        super(utils_1.runtimeRoot('package.json'));
        if (this.model !== null) {
            this.name = this.model.name;
            this.version = this.model.version;
            this.description = this.model.description;
            this.repository = this.model.repository;
            this.keywords = this.model.keywords;
            this.author = this.model.author;
            this.license = this.model.license;
            this.homepage = this.model.homepage;
            this.bugs = this.model.bugs;
            this.scripts = this.model.scripts;
            this.dependencies = this.model.dependencies;
            this.devDependencies = this.model.devDependencies;
            this.jest = this.model.jest;
        }
    }
    updateModuleNameMapper() {
        const aliases = VuesionConfig_1.VuesionConfig.getWebpackAliases();
        if (aliases) {
            Object.keys(aliases).map((alias) => {
                this.model.jest.moduleNameMapper[`^${alias}/(.*)$`] = `<rootDir>/${aliases[alias]}/$1`;
            });
            this.save(false);
        }
    }
    save(prettier = true) {
        if (this.model !== null) {
            this.model.name = this.name;
            this.model.version = this.version;
            this.model.description = this.description;
            this.model.repository = this.repository;
            this.model.keywords = this.keywords;
            this.model.author = this.author;
            this.model.license = this.license;
            this.model.homepage = this.homepage;
            this.model.bugs = this.bugs;
            this.model.scripts = this.scripts;
            this.model.dependencies = this.dependencies;
            this.model.devDependencies = this.devDependencies;
            this.model.jest = this.jest;
        }
        super.save(prettier);
    }
}
exports.VuesionPackage = new Model();
