// @vuesion/service/dist/generators/crud-module/index.js
"use strict";
const path = require("path");
const ast_1 = require("../ast");
const models_1 = require("@vuesion/models");
const path_1 = require("@vuesion/utils/dist/path");
const fileSystem_1 = require("@vuesion/utils/dist/fileSystem");
const pluralize = require('pluralize');
module.exports = {
    description: 'Add a module with CRUD operations (POST, GET, PUT, DELETE)',
    prompts: [
        {
            type: 'input',
            name: 'name',
            message: "What's the name of your API endpoint? For example: /posts",
            validate: (value) => {
                if (!value || value.length === 0) {
                    return 'name is required';
                }
                value = value.slice(-1).toLocaleLowerCase() === 's' ? value.slice(0, -1) : value;
                return fileSystem_1.folderExists(path_1.runtimeRoot(path.join(models_1.VuesionConfig.generators.outputDirectory, value)))
                    ? `folder already exists (${value})`
                    : true;
            },
        },
    ],
    actions: (data) => {
        const filePath = data.name.split('/').filter((part) => part.trim().length > 0);
        const moduleName = filePath.pop();
        const singularName = moduleName.slice(-1).toLocaleLowerCase() === 's' ? moduleName.slice(0, -1) : moduleName;
        const pluralName = pluralize(singularName);
        data.moduleName = singularName;
        data.singularName = singularName;
        data.pluralName = pluralName;
        data.modulePath = filePath.join('/');
        data.componentName = data.singularName;
        data.basePath = path.join(process.cwd(), models_1.VuesionConfig.generators.outputDirectory, filePath.join('/'));
        data.wantRoutes = true;
        data.wantVuex = true;
        let actions = [
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/{{properCase componentName}}/{{properCase componentName}}.vue',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'connected/connected.vue.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/{{properCase componentName}}/{{properCase componentName}}.spec.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'connected/connected.spec.ts.hbs'),
                abortOnFail: true,
            },
        ];
        actions.push({
            type: 'add',
            path: '{{basePath}}/{{camelCase singularName}}/routes.ts',
            templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/routes.ts.hbs'),
            abortOnFail: true,
        });
        ast_1.addModuleToRouter(path.join(path.resolve(process.cwd()), models_1.VuesionConfig.generators.routerFile), data.singularName, data.modulePath);
        actions = actions.concat([
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/actions.spec.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/actions.spec.ts.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/actions.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/actions.ts.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/getters.spec.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/getters.spec.ts.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/getters.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/getters.ts.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/module.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/module.ts.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/mutations.spec.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/mutations.spec.ts.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/mutations.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/mutations.ts.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/state.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/state.ts.hbs'),
                abortOnFail: true,
            },
            {
                type: 'add',
                path: '{{basePath}}/{{camelCase singularName}}/I{{properCase singularName}}.ts',
                templateFile: path.join(process.cwd(), models_1.VuesionConfig.generators.blueprintDirectory, 'crud-module/interface.ts.hbs'),
                abortOnFail: true,
            },
        ]);
        ast_1.addModuleToState(path.join(path.resolve(process.cwd()), models_1.VuesionConfig.generators.stateFile), data.singularName, data.modulePath);
        return actions;
    },
};
