// @embarklabs/subspace/lib/database/nullDatabase.js
import _classCallCheck from "@babel/runtime-corejs3/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/esm/createClass";

var NullDatabase =
/*#__PURE__*/
function () {
  function NullDatabase(_dbFilename, events, cb) {
    _classCallCheck(this, NullDatabase);

    this.events = events;

    if (cb) {
      cb();
    }
  }

  _createClass(NullDatabase, [{
    key: "databaseInitialize",
    value: function databaseInitialize(cb) {
      if (cb) {
        cb();
      }
    }
  }, {
    key: "getLastKnownEvent",
    value: function getLastKnownEvent() {
      return {
        firstKnownBlock: 0,
        lastKnownBlock: 0
      };
    }
  }, {
    key: "getEventsFor",
    value: function getEventsFor(eventKey) {
      return [];
    }
  }, {
    key: "eventExists",
    value: function eventExists(eventKey, eventId) {
      return false;
    }
  }, {
    key: "recordEvent",
    value: function recordEvent(eventKey, values) {}
  }, {
    key: "deleteEvent",
    value: function deleteEvent(eventKey, eventId) {}
  }, {
    key: "deleteNewestBlocks",
    value: function deleteNewestBlocks(eventKey, gteBlockNum) {}
  }]);

  return NullDatabase;
}();

export default NullDatabase;
